src/grav/test.py

```diff
@@ -0,0 +1,115 @@
+"""
+This file contains meat of the Gravitational QA coding challenge.
+
+The test() function performs the test as requested in the challenge.
+
+The setup() function does all the provisioning necessary to go from:
+  "package is installed" -> "start/stop/logs a wellknown container"
+
+The teardown() function attempts to unwind any dangling resources
+from setup() and test(). Ideally leaving a clean test enviromnent
+if it runs to completion.  This is important for idempotency.
+"""
+import re
+from pkg_resources import resource_filename
+
+from grav.client import demux_logs, DockerClient
+
+
+def setup():
+    # If this were real, I'd pass these magic strings stattered through
+    # file as params and have them loaded from a config.
+    image_tar = resource_filename(__name__, "wellknown.tar")
+
+    # It'd be better to nest client, image, container setup/teardown
+    # so that in the case client & image succeed but container fails
+    # we only run the teardown for image & client, and skip container
+    # teardown since it was never set up.
+    #
+    # That is more effort than this sample merits imo, so setup/teardown
+    # are all in one. -- wdella 2019-10
+    client = DockerClient("/var/run/docker.sock")
+
+    # load image from tar, get image id
+    with open(image_tar, 'rb') as fh:
+        data = fh.read()
+
+    headers = {'Content-Type': 'application/tar', 'Content-Length': len(data)}
+    resp = client.post("/images/load?fromSrc=-", body=data, headers=headers)
+    assert resp.status in (200, 201)
+    json = resp.json()
+    match = re.match(r"Loaded image ID: (sha256:[\da-f]{64})", json['stream'])
```

Comment:
is there a way around parsing unstructured data?

Reply:
Not that I've found.  The docs on this are silent about expected output:

https://docs.docker.com/engine/api/v1.30/#operation/ImageCreate

And I haven't found a way to force a more properly structured json response than the `{"stream": "Loaded image ID: ..."}` that you see here.  If you or anyone else knows of a way around this, I'd be thrilled to cut the regex.